Fix parser bug that prevents concurrent parsing.

LGM-Doyle opened the pull request:
PR #1785 depends on this PR.

This PR contains all of the commits for PRs #1797, #1801, #1802, and #1803 which were created to track down a pair of declarations that causes the MSVC Appveyor CI to fail.

This PR contains all of the fixes to the parser that allow it to correctly concurrently parse files.  It will also enable multi-threaded parsing, which is the goal of #1785. 

It parses the same as the current master branch.

The problem as described in more detail in #1785 is that the lexer is a static singleton that can only be bound to a single file/string for parsing at one time.  

The fix is to make lexers that are non-static, and non-singleton and bind one to each file.

In order to do that this PR finds each singleton parser that is bound to the lexer,  converts it to a grammar, passes it as a parameter to its dependents and then removes the singleton.

LGM-Doyle committed:
Start removing lexer singleton.

The lexer is a singleton.

The lexer is bound to the file being parsed and maintains the state of
the tokens being parsed.

Because the lexer is a singleton only one file can be parsed in any
executable at one time.  This means that there can be no recursive
parsing (macros that are not pre-parsed) nor any multi-threaded parsing.

This commit begins removing the lexer singleton by making its
constructor public.

LGM-Doyle committed:
Remove references to lexer singleton from ValueRefParserImpl.cpp

LGM-Doyle committed:
Remove double_parser_rules() singleton refs.

The double_parser_rules singleton includes member variables that are
referenced.  When it is promoted to a member variable it means that its
member types are promoted to public member types.

LGM-Doyle committed:
Remove int_arithmetic_rules() singleton refs.

The int_arithmetic_rules singleton includes member variables that are
referenced.  When it is promoted to a member variable it means that its
member types are promoted to public member types.

LGM-Doyle committed:
Remove rest of int parser singletons from IntValueRefParser.cpp

The int parser singletons include references to the lexer singleton and
export references to their member variables. All these references
prevent multiple files from being parsed at the same time.

LGM-Doyle committed:
Change singleton refs to parse::lexer into parameters in DoubleValueParser

LGM-Doyle committed:
Change singleton refs to parse::lexer into parameters in IntValueRefParser

LGM-Doyle committed:
In ParseImpl.h pass lexer as parameter to rules.

This is a step to remove lexer as a singleton and allow concurrent
parsing.

LGM-Doyle committed:
Change lexer::m_name_tokens to unordered_map.

name_tokens are never listed alphabetically.

LGM-Doyle committed:
Create parse::detail::Labeller to replace parse::detail::label()

parse::detail::label() contains the embedded singleton
static std::map<...> rules;
Furthermore, label() is implicitly bound to the lexer::instance()
singleton. Both causes prevent labels being generated correctly on
different threads.

This commit creates Labeller which will replace uses of label.  Labeller
is explicitly bound to a particular instance of lexer and does not
contain a singleton.

LGM-Doyle committed:
Replace some uses of parse::detail::label() with parse::detail::Labeller()

Pass a Labeller instance bound to the lexer created in parse_file() as a
parameter of the constructor of the Rules.  Use this labeller to replace
top level uses of label().

LGM-Doyle committed:
Add parse::detail::common_params_rules

common_params_rules allows rules to create the common_params_rules
locally to allow for concurrent parsing.

LGM-Doyle committed:
Use parse::detail::common_params_rules.

LGM-Doyle committed:
Remove unused CommonParamsParser singleton.

LGM-Doyle committed:
Refactor EffectParser 4 and 5 with direct reference to parse::effect_parser

Currently EffectParser 4 and 5 refer to the singleton
parse::effect_parser, which prevents the effect parser from being made
into a grammar with a normal constructor.

LGM-Doyle committed:
Add parse::grammar

boost::spirit::qi::grammar models a collection of rule with a
distinguished start/base rule.  Many rules in freeorion have re-invented
this type and can be simplified by using the spirit::qi provided
mechanisms.

LGM-Doyle committed:
Replace parse::effect_parser() with effects_grammar

The effects_grammar is a boost::spirit::qi::grammar that does not rely
on the singleton parse::detail::effect_parser.

LGM-Doyle committed:
Remove unused effect_parser singleton.

LGM-Doyle committed:
Move parse:detail::effects_group_parser into EffectParser.cpp

LGM-Doyle committed:
Change parse::effects_group_parser into a grammar.

Since the grammar knows its own start rule, there is no need for a an
object that persists longer than the rule.  The effects_group_parser
singleton is also removed.

LGM-Doyle committed:
Pass the labeller to effects_group_grammar as a parameter.

LGM-Doyle committed:
Pass labeller as parameter to EffectParser 1 thru 5.

LGM-Doyle committed:
Add int type to some enums in Condition.h

This will allow these enums to be forward declared.

LGM-Doyle committed:
Refactor ConditionParser 1, 3 and 7 with direct reference to parse::condition_parser

Currently ConditionParser 1, 3 and 7 refer to the singleton
parse::condition_parser, which prevents the condition parser from being
made into a grammar with a normal constructor.

LGM-Doyle committed:
Convert refs to parse::condition_parser singleton to params.

LGM-Doyle committed:
Change parse::condition_parser into a grammar.

Since the grammar knows its own start rule, there is no need for an
object that persists longer than the rule.

LGM-Doyle committed:
Add condition_parser parameters to some cond/effect parser parts.

These components will need the condition_parser as parameters to the int
and double parsers for the statistical portions.

LGM-Doyle committed:
Refactor arithmetic_rules to use condition_parser param.

This converts arithmetic_rules to not use the singleton
condition_parser, which means arithmetic_rules can be used concurrently
in multiple parsers.

LGM-Doyle committed:
Remove the double_complex_parser singleton.

LGM-Doyle committed:
Remove the int_complex_parser singleton.

LGM-Doyle committed:
Remove unused double_complex_parser_rules and int_complex_parser_rules.

LGM-Doyle committed:
Add using statement for parse::value_ref_grammar.

LGM-Doyle committed:
Refactor string_parser_rules singleton into a passed parameter.

LGM-Doyle committed:
Move declaration of enum_value_ref<T> to ValueRefParser.h

This allows various enum_value_ref_rules<T> singleton creation functions
to be converted to structs.

LGM-Doyle committed:
Change parse::detail::universe_object_type from singleton to parameter.

LGM-Doyle committed:
Change parse::detail::star_type_rules from singleton to parameter.

LGM-Doyle committed:
Change parse::detail::planet_type_rules from singleton to parameter.

LGM-Doyle committed:
Change parse::detail::planet_size_rules from singleton to parameter.

LGM-Doyle committed:
Change parse::detail::planet_environment_rules from singleton to parameter.

LGM-Doyle committed:
Change enum_value_ref_rules<T>() to take refs to tok and condition_parser

LGM-Doyle committed:
Change initialize_nonnumeric_statistic_parser() to take refs to tok and condition_parser

LGM-Doyle committed:
Replace condition_parser() singleton with condition_parser_grammar

LGM-Doyle committed:
Add tags_grammar, to replace tags_parser() singleton.

LGM-Doyle committed:
Replace tags_parser() singleton with tags_grammar.

LGM-Doyle committed:
Replace the item_spec_parser singleton.

LGM-Doyle committed:
Replace string_var_complex singleton.

LGM-Doyle committed:
Replace parse::detail::label() singleton

LGM-Doyle committed:
Replace various singletons from EnumParser.h

LGM-Doyle committed:
Replace parse::color_parser() singleton

LGM-Doyle committed:
Remove lexer singleton refs in simple_variable_rules

LGM-Doyle committed:
Remove lexer singleton refs in EffectParser5.cpp

LGM-Doyle committed:
Remove unused parse::label() definition.

LGM-Doyle committed:
Change parse::double_ and parse::int_ singletons to grammars.

LGM-Doyle committed:
Remove unused parse::init()

LGM-Doyle committed:
Make DoubleComplexValueRefParser string non static.

static string may/may not be initialized at the same time in multiple
threads.

LGM-Doyle committed:
Remove the parse::lexer::instance() singleton.

LGM-Doyle committed:
Change labeller to a variable in each parser.

The labeller only requires the lexer as a parameter to be constructed.
There is no reason to pass both as parameters to the parsing
rule/grammar.

LGM-Doyle committed:
Change parse::parse_file to use a grammar.

Previously, it used a set of rules with a special rule called start,
which is the spirit::grammar use case.

LGM-Doyle committed:
Make TestEnumParser work with non-singleton enums.

LGM-Doyle committed:
Remove unused parse::ship_designs and parse::monster_designs

LGM-Doyle committed:
Reuse lexer as param to parse_file.

This change minimizes the overhead from initialization of the lexer
deterministic finite automaton (DFA).

LGM-Doyle committed:
Change parse::detail::visibility_rules from singleton to parameter.

LGM-Doyle committed:
Replace the visibility_var_complex singleton.

LGM-Doyle committed:
Replace parse_file timer with a ScopedTimer

The scoped timer as a 100us threshold and can be turned off by
players/testers who do not need timing information.

LGM-Doyle committed:
Restore file length checking and add error reporting to parse_file().

LGM-Doyle committed:
Remove duplicate definition of parse::grammar from EnumParser.h

LGM-Doyle committed:
Simplify includes in Condition and Effect parsers.

LGM-Doyle committed:
Further simplify includes in EffectParser.h

dbenage-cx reviewed (commented):
Couple of formatting notes.

LGM-Doyle committed:
Break EffectParser.h into EffectParser[1..5].h

This will speed up re-compilation when changes only affect a single
portion of the EffectParser.

LGM-Doyle committed:
Break CondtionParser.h into ConditionParser[1..7].h

This will speed up re-compilation when changes only affect a single
portion of the CondtionParser.

LGM-Doyle committed:
Limit unnecessary use of parse:: namespace when instantiating lexer.

LGM-Doyle committed:
Reduce unnecessary use of parse::detail namespace.

Issue raised in review of PR #1808.

LGM-Doyle committed:
Fix indentation in ValueRefParser.h

Issue raised in review of PR #1808.

LGM-Doyle committed:
housekeeping: deindent effects_parser_grammar body

for PR review #1808.

LGM-Doyle reviewed (commented):
Thanks for reviewing.

I've made the changes.

I moved two commits that restore the end of file character counting from #1785 to this PR, since I had implied they were already included.  The changes are only in "ParseImpl.h"  `parse_file()`.

LGM-Doyle committed:
housekeeping: style changes for review #1808

LGM-Doyle reviewed (commented):
The project could adopt a C++ style checker. I've configured and run a few on the code base.

The freeorion C++ style is idiosyncratic.  None of the style checkers I've tried can be configured match its unique style.

The better maintained style checkers run on many editors on all 3 of our supported platforms and/or have easily configured git pre-commit and pre-push hooks.   However, they are less configurable.  The consensus is that with fewer options it is easier to guarantee the source code transformation does not introduce bugs.

Adopting a style checker, would mean there would no longer be a style component to the review process.
